zebra-scan/src/bin/scanner/main.rs

```diff
@@ -0,0 +1,137 @@
+use lazy_static::lazy_static;
+use structopt::StructOpt;
+use tracing::*;
+
+use zebra_chain::{block::Height, parameters::Network};
+use zebra_state::{ChainTipSender, SaplingScanningKey};
+
+use core::net::SocketAddr;
+use std::path::PathBuf;
+
+/// A strucure with sapling key and birthday height.
+#[derive(Clone, Debug, Eq, PartialEq, serde::Deserialize)]
+pub struct SaplingKey {
+    key: SaplingScanningKey,
+    #[serde(default = "min_height")]
+    birthday_height: Height,
+}
+
+fn min_height() -> Height {
+    Height(0)
+}
+
+impl std::str::FromStr for SaplingKey {
+    type Err = Box<dyn std::error::Error>;
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        Ok(serde_json::from_str(value)?)
+    }
+}
+
+#[tokio::main]
+/// Runs the zebra scanner binary with the given arguments.
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Display all logs from the zebra-scan crate.
+    tracing_subscriber::fmt::fmt()
+        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+        .init();
+
+    // Parse command line arguments.
+    let args = Args::from_args();
+
+    let zebrad_cache_dir = args.zebrad_cache_dir;
+    let scanning_cache_dir = args.scanning_cache_dir;
+    let mut db_config = zebra_scan::Config::default().db_config;
+    db_config.cache_dir = scanning_cache_dir;
+    let network = args.network;
+    let sapling_keys_to_scan = args
+        .sapling_keys_to_scan
+        .into_iter()
+        .map(|key| (key.key, key.birthday_height.0))
+        .collect();
+    let listen_addr = args.listen_addr;
+
+    // Create a state config with arguments.
+    let state_config = zebra_state::Config {
+        cache_dir: zebrad_cache_dir,
+        ..zebra_state::Config::default()
+    };
+
+    // Create a scanner config with arguments.
+    let scanner_config = zebra_scan::Config {
+        sapling_keys_to_scan,
+        listen_addr,
+        db_config,
+    };
+
+    // Get a read-only state and the database.
+    let (read_state, db, _) = zebra_state::init_read_only(state_config, &network);
+
+    // Get the initial tip block from the database.
+    let initial_tip = db
+        .tip_block()
+        .map(zebra_state::CheckpointVerifiedBlock::from)
+        .map(zebra_state::ChainTipBlock::from);
+
+    // Create a chain tip sender and use it to get a chain tip change.
+    let (_chain_tip_sender, _latest_chain_tip, chain_tip_change) =
+        ChainTipSender::new(initial_tip, &network);
+
+    // Spawn the scan task.
+    let scan_task_handle =
+        { zebra_scan::spawn_init(scanner_config, network, read_state, chain_tip_change) };
+
+    // Pin the scan task handle.
+    tokio::pin!(scan_task_handle);
+
+    // Wait for task to finish
+    loop {
+        let _result = tokio::select! {
+            scan_result = &mut scan_task_handle => scan_result
+                .expect("unexpected panic in the scan task")
+                .map(|_| info!("scan task exited")),
+        };
+    }
```

Comment:
I think we want to break when the scan task finishes, if it's running without a gRPC server, it won't be adding any more results to the scanner db, if it's running with a server, the scan task shouldn't finish unless the server shut down.

Optional: Could we rebase this on https://github.com/ZcashFoundation/zebra/pull/8596 and address https://github.com/ZcashFoundation/zebra/issues/8633 in this PR as well?

If you do the rebase, you could merge the last commit [on this `scanner-with-rpc-syncer`](https://github.com/ZcashFoundation/zebra/compare/scanner-with-rpc-syncer) branch into this PR if it looks good.

Reply:
thanks, this was done as suggested.